torch/csrc/jit/codegen/cuda/lower_loops.cpp

```diff
@@ -303,15 +304,234 @@ void LoopNestGenerator::handle(Expr* expr) {
+// Sort each loop-nest group based on axis (i.e., score)
+void sortGroup(TensorView* target, ExprListT& exprs, ExprScoreMapT& scores) {
+  std::stable_sort(
+      exprs.begin(),
+      exprs.end(),
+      [&scores](const Expr* expr1, const Expr* expr2) {
+        return scores[expr1] < scores[expr2];
+      });
+}
+
+void mergeNonRootGroupsIntoRootGroups(
+    TargetGroupMapT& computed_at_exprs,
+    ExprTargetMapT& target_map) {
+  for (auto it = computed_at_exprs.begin(); it != computed_at_exprs.end();) {
+    TensorView* target = it->first;
+    if (target->hasComputeAt()) {
+      Expr* target_expr = target->getOrigin();
+      TensorView* target_of_target = target_map.at(target_expr);
+      auto& target_group = computed_at_exprs.at(target_of_target);
+      auto pos =
+          std::find(target_group.begin(), target_group.end(), target_expr);
+      TORCH_INTERNAL_ASSERT(pos != target_group.end());
+      target_group.insert(pos, it->second.begin(), it->second.end());
+      // Upate the target map
+      for (auto& inserted_expr : it->second) {
+        TORCH_INTERNAL_ASSERT(target_map.at(inserted_expr) == target);
+        target_map.at(inserted_expr) = target_of_target;
+      }
+      it = computed_at_exprs.erase(it);
+    } else {
+      ++it;
+    }
+  }
+}
+
+// Merge root loop-nests into reordered_exprs
+void mergeGroupsIntoSortedList(
+    TargetGroupMapT& computed_at_exprs,
+    ExprListT& reordered_exprs) {
+  while (computed_at_exprs.size() > 0) {
+    // Find the root loop-nest that has no dependency with the other
+    // loop-nests
+    TensorView* cur_target = computed_at_exprs.begin()->first;
+    for (auto& group : computed_at_exprs) {
+      auto target = group.first;
+      if (cur_target == target)
+        continue;
+      if (DependencyCheck::isDependencyOf(target, cur_target)) {
+        cur_target = target;
+      }
+    }
+    // cur_target can be visited
+    reordered_exprs.insert(
+        reordered_exprs.end(),
+        computed_at_exprs.at(cur_target).begin(),
+        computed_at_exprs.at(cur_target).end());
+    computed_at_exprs.erase(cur_target);
+  }
+}
+
+// Reorder exprs so that LoopNestGenerator::handle(Expr*) can generate
+// correct loop nests. Vector exprs is assumed to be topologically
+// sorted, but that is not sufficient as tensors computed at
+// outer loops need to be located earlier.
+void reorderExprsForComputeAt(std::vector<Expr*>& exprs) {
+  ExprListT reordered_exprs;
+  // expr -> target
+  ExprTargetMapT target_map;
+  // target -> [computed at expressions]
+  TargetGroupMapT computed_at_exprs;
+  // score of each expression that is calculated based on the
+  // computeAt axis. A lower score of an expression means it should be
+  // placed earlier in the expression list. This is a requirement for
+  // the loop-nest generation of this class to work.
+  ExprScoreMapT scores;
+
+  // 1. Group expressions by target tensors. Non-grouped expressions
+  // are copied into reordered_exprs.
```

Comment:
Is a target tensor a tensor with no computeAt?

Reply:
Not necessarily. Please see the manual.